src/liballoc/rc.rs

```diff
@@ -1708,9 +1702,18 @@ impl<T> Weak<T> {
     /// [`null`]: ../../std/ptr/fn.null.html
     #[stable(feature = "weak_into_raw", since = "1.45.0")]
     pub fn as_ptr(&self) -> *const T {
-        let offset = data_offset_sized::<T>();
-        let ptr = self.ptr.cast::<u8>().as_ptr().wrapping_offset(offset);
-        ptr as *const T
+        let ptr: *mut RcBox<T> = NonNull::as_ptr(self.ptr);
+        let fake_ptr = ptr as *mut T;
+
+        // SAFETY: we must offset the pointer manually, and said pointer may be
+        // a dangling weak (usize::MAX). data_offset is safe to call, because we
+        // know a pointer to unsized T must be derived from a real unsized T,
+        // because dangling weaks are only created for sized T. wrapping_offset
+        // is used so that we can use the same code path for dangling weak refs.
```

Comment:
```suggestion
        // is used so that we can use the same code path for the non-dangling unsized
        // case and the dangling sized case.
```

Reply:
Reworded this slightly to point say "potentially dangling sized case", but did apply the suggestions manually.